src/cmap/mod.rs

```diff
@@ -388,14 +293,84 @@ impl ConnectionPoolInner {
             return Ok(conn);
         }
 
+        // drop the lock before we create a new connection to allow other connections to stop
+        // waiting.
+        drop(available_connections_lock);
+
         // There are no connections in the pool, so open a new one.
-        let connection = connection_manager.create_connection().await?;
-        wait_queue_handle.disarm();
-        Ok(connection)
+        let pending_connection = self.create_pending_connection();
```

Comment:
I'm not sure it really matters, so I just opted to drop the lock as soon as possible. Did you have a case in mind where creating it after would cause issues?

Reply:
If the lock is dropped beforehand, theoretically maxPoolSize could be reached between when the lock is dropped and when the pending connection is created, leading us to exceed maxPoolSize. More generally, changing the total connection count without holding the lock will potentially be susceptible to races.